File: 2025/python/benchmark_day_01.py
#!/usr/bin/env python3
from timeit import timeit
from pathlib import Path
from typing import Iterator, Tuple
from Functions_2025 import read_lines

INPUT_PATH = Path("Advent_of_Code_2025/input_files") / "Password_File.txt"

def get_rotation_params(line: str) -> Tuple[str, int]:
    direction = line[0]
    rotation_value = int(line[1:].strip())
    return direction, rotation_value

# WRAPPER FUNCTIONS - timeit SAFE
def benchmark_part1():
    rotations = (get_rotation_params(line) for line in read_lines(INPUT_PATH))
    dial_position = 50
    count = 0
    for direction, rotation_value in rotations:
        rotation_angle = rotation_value if direction == 'R' else -rotation_value
        dial_position = (dial_position + rotation_angle) % 100
        if dial_position == 0:
            count += 1
    return count

def benchmark_part2():
    rotations = (get_rotation_params(line) for line in read_lines(INPUT_PATH))
    dial_position = 50
    count = 0
    for direction, rotation_value in rotations:
        rotation_angle = rotation_value if direction == 'R' else -rotation_value
        dial_position += rotation_angle
        if dial_position < 0:
            if dial_position % 100 == 0:
                count += abs(dial_position) // 100
            else:
                count += abs(dial_position) // 100 + 1
        elif dial_position >= 100:
            count += dial_position // 100
        dial_position %= 100
    return count

# BENCHMARKS - No external vars!
number_of_runs = 10000
print(f"=== PART 1 BENCHMARK ({number_of_runs} runs) ===")
t1 = timeit('benchmark_part1()', number=number_of_runs, globals=globals())
print(f"Avg: {t1/number_of_runs:.7f}ms")

print("\n=== PART 2 BENCHMARK (1,000 runs) ===")
t2 = timeit('benchmark_part2()', number=number_of_runs, globals=globals())
print(f"Avg: {t2/number_of_runs:.7f}ms")

print(f"\nP1/P2 ratio: {t1/t2:.2f}x")

File: 2025/cpp/benchmark_day_01.cpp
#include "Functions_2025.h"
#include <iostream>


int part1(const std::vector<Rotation>& rotations){

    int dial_position = 50;
    int count = 0;

    for (const auto& r: rotations){
        int rotation_angle = (r.direction == 'R') ? r.rotation_value : -r.rotation_value;
        // trenary operator: condition ? true_value: false_value
        dial_position = (dial_position + rotation_angle ) % 100;
        if (dial_position == 0){
            count++;
        }
    }
    return count;
}


int part2(const std::vector<Rotation>& rotations) {
    int dial_position = 50;
    int count = 0;
    
    for (const auto& r : rotations) {
        int rotation_angle = (r.direction == 'R') ? r.rotation_value : -r.rotation_value;
        dial_position += rotation_angle;
        
        if (dial_position < 0) {
            if (dial_position % 100 == 0) {
                count += std::abs(dial_position) / 100;
            } else {
                count += std::abs(dial_position) / 100 + 1;
            }
        } else if (dial_position >= 100) {
            count += dial_position / 100;
        }
        dial_position = mod(dial_position, 100);
    }
    return count;
}

double benchmark(int N, auto&& func) {
    double total = 0;
    for (int i = 0; i < N; ++i) {
        auto start = std::chrono::high_resolution_clock::now();
        func();
        auto end = std::chrono::high_resolution_clock::now();
        total += std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0;
    }
    return total / N;
}

int main() {
    auto lines = read_lines("../input_files/Password_File.txt");
    std::vector<Rotation> rotations;
    for (const auto& line : lines) {
        rotations.push_back(parse_line(line));
    }
    
    const int N = 10000;  // 10k runs
    std::cout << "Benchmarking " << N << " runs (average):\n";
    
    double parse_avg = benchmark(N, [&]() {
        std::vector<Rotation> rotations;
        for (const auto& line : lines) rotations.push_back(parse_line(line));
    });
    
    double p1_avg = benchmark(N, [&]() { part1(rotations); });
    double p2_avg = benchmark(N, [&]() { part2(rotations); });
    
    std::cout << "Parse: " << parse_avg << " ms (x" << N << ")\n";
    std::cout << "Part 1: " << part1(rotations) << " (" << p1_avg << " ms)\n";
    std::cout << "Part 2: " << part2(rotations) << " (" << p2_avg << " ms)\n";
}

